optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/PlannerBenchmarkRunner.java

```diff
@@ -122,54 +127,134 @@ private void initBenchmarkDirectoryAndSubdirs() {
+        // Wait for the warm up benchmarks to complete
+        int tasksCount = futureMap.size();
+        for (int i = 0; i < tasksCount; i++) {
+            Future<SingleBenchmarkRunner> future;
             try {
-                Solver solver = solverBenchmarkResult.getSolverConfig().buildSolver();
-                solver.solve(problemBenchmarkResult.readPlanningProblem());
-            } catch (RuntimeException e) {
-                singleBenchmarkRunnerExceptionLogger.warn("The warmUp of singleBenchmark ("
-                        + singleBenchmarkResult.getName() + ") failed.", e);
-                // TODO update firstFailureSingleBenchmarkRunner (when warmups happen in a Runner too in parallel)
+                future = warmUpExecutorCompletionService.take();
+            } catch (InterruptedException e) {
+                singleBenchmarkRunnerExceptionLogger.error("Waiting for a warm up singleBenchmarkRunner was interrupted.", e);
+                continue;
             }
 
-            solverBenchmarkResult.getSolverConfig().setTerminationConfig(originalTerminationConfig);
-            long timeSpent = System.currentTimeMillis() - startingTimeMillis;
-            timeLeft = warmUpTimeMillisSpentLimit - timeSpent;
-            if (timeLeft <= 0L) {
-                return timeLeft;
+            Throwable failureThrowable = null;
+            SingleBenchmarkRunner singleBenchmarkRunner = futureMap.get(future);
+            try {
+                // Explicitly returning it in the Callable guarantees memory visibility
+                singleBenchmarkRunner = future.get();
+            } catch (InterruptedException e) {
+                singleBenchmarkRunnerExceptionLogger.error("The warm up singleBenchmarkRunner ({}) was interrupted.",
+                        singleBenchmarkRunner, e);
+                failureThrowable = e;
+            } catch (ExecutionException e) {
+                Throwable cause = e.getCause();
+                singleBenchmarkRunnerExceptionLogger.warn("The warm up singleBenchmarkRunner ({}) failed.",
+                        singleBenchmarkRunner, cause);
+                failureThrowable = cause;
+            }
+            if (failureThrowable != null) {
+                singleBenchmarkRunner.setFailureThrowable(failureThrowable);
+                if (firstFailureSingleBenchmarkRunner == null) {
+                    firstFailureSingleBenchmarkRunner = singleBenchmarkRunner;
+                }
+            }
+
+            SolverBenchmarkResult solverBenchmarkResult = singleBenchmarkRunner.getSingleBenchmarkResult().getSolverBenchmarkResult();
+            long timeLeftInCycle = timePhaseEnd - System.currentTimeMillis();
+            if (timeLeftInCycle > 0L) {
+                warmUpPopulate(futureMap, singleBenchmarkResultIndexMap, new SolverBenchmarkResult[]{solverBenchmarkResult}, timeLeftInCycle);
+                tasksCount++;
```

Comment:
Aggreed on tasksCounts++.

I still dislike the trick of using the tasksCounts to predict when the queue will be empty and the loop should stop. See my comment on poll() and take() above.
BTW: the reason there is no isEmpty() method in that service is for multi-threaded use cases (even though we only have 1 consumer thread): the poll() method does the isEmpty() check inherently in a lock safe manner.